File: src/entities/objects/flame.ts
import { GameObject } from '@/entities/base';
import { COLOR_FLAME } from '@/game/constants';

/**
 * Flame object - needs to be extinguished to win
 */
export class Flame extends GameObject {
  private brightness: number = 1.0;
  private flickerTime: number = 0;

  constructor(x: number = 0, y: number = 0) {
    super(x, y);
    this.setProperty('solid', false); // Flames don't block movement
    this.setProperty('pushable', false);
    this.setProperty('fragile', true); // Can be extinguished
    this.setProperty('supports_weight', false);
    this.setProperty('weight', 0);
    this.setProperty('is_fire', true);
  }

  getType(): string {
    return 'flame';
  }

  getColor(): number {
    return COLOR_FLAME;
  }

  update(dt: number): void {
    // Flicker animation
    this.flickerTime += dt * 3; // 3 flickers per second
    this.brightness = 0.7 + Math.sin(this.flickerTime) * 0.3;
  }

  onCollision(other: GameObject): void {
    // Check if colliding with ice block
    if (other.getType() === 'ice_block') {
      this.extinguish();
    }
    
    // Check if colliding with ice pot
    if (other.getType() === 'pot' && other.getProperty('is_cold')) {
      this.extinguish();
    }
  }

  private extinguish(): void {
    this.destroy();
  }

  getBrightness(): number {
    return this.brightness;
  }

  canIgnite(): boolean {
    return true; // Flames can ignite other objects
  }

  isFirm(): boolean {
    return false; // Flames are not firm
  }
}

File: src/entities/objects/flame.py
# Flame Class

from src.entities.base import GameObject
from src.game.constants import COLOR_FLAME


class Flame(GameObject):
    """Flame object that needs to be extinguished"""
    
    def __init__(self, x: int = 0, y: int = 0):
        super().__init__(x, y)
        
        # Flame properties
        self.set_property('solid', True)
        self.set_property('pushable', False)
        self.set_property('fragile', True)  # Can be extinguished by ice
        self.set_property('affected_by_gravity', False)
        self.set_property('supports_weight', True)
        self.set_property('interactive', False)
        self.set_property('height', 1)
        self.set_property('weight', 0)
        
        # Flame-specific properties
        self.set_property('burning', True)
        self.set_property('can_extinguish_ice_above', False)  # Ice above doesn't melt
    
    def get_type(self) -> str:
        return "flame"
    
    def get_color(self) -> tuple:
        return COLOR_FLAME
    
    def update(self, dt: float):
        """Update flame animation"""
        super().update(dt)
        # TODO: Add flame animation
    
    def on_collision(self, other: GameObject):
        """Handle collision with other objects"""
        if other.get_type() == "ice_block":
            # Ice extinguishes flame
            self.destroy()
    
    def is_burning(self) -> bool:
        """Check if flame is still burning"""
        return self.get_property('burning', True)
    
    def extinguish(self):
        """Extinguish the flame"""
        self.set_property('burning', False)
        self.destroy()
    
    def can_melt_ice_above(self) -> bool:
        """Check if flame can melt ice directly above"""
        return self.get_property('can_extinguish_ice_above', False)